File: examples/languages/c/leap-year/leap.c
// Volatile sink forces the load — without this, clang at -O1
// constant-folds the predicate result and emits zero branches.
volatile unsigned year_input;

static int leap_year(unsigned y) {
    return (y % 4 == 0 && y % 100 != 0) || (y % 400 == 0);
}

__attribute__((export_name("run_row_0")))
int run_row_0(void) { year_input = 2001; return leap_year(year_input); }

__attribute__((export_name("run_row_1")))
int run_row_1(void) { year_input = 2004; return leap_year(year_input); }

__attribute__((export_name("run_row_2")))
int run_row_2(void) { year_input = 2100; return leap_year(year_input); }

__attribute__((export_name("run_row_3")))
int run_row_3(void) { year_input = 2000; return leap_year(year_input); }

File: examples/languages/zig/leap-year/leap.zig
// Zig — leap-year MC/DC probe.
//
// Same predicate as the C/Rust fixtures: `(y%4==0 and y%100!=0)
// or (y%400==0)`. Zig's `and`/`or` short-circuit operators lower
// to LLVM IR essentially identical to clang's, so the wasm
// emission should be `if/else` + 1 `br_if` per source decision —
// the shape v0.19's IfThen clustering targets.
//
// We use a `noinline` predicate + an exported `year_input` global
// the runner mutates between calls — that defeats Zig's constant
// folding without dragging in atomics (whose memory-ordering
// machinery dominated the branch counts in earlier iterations).

export var year_input: u32 = 0;

noinline fn leap_year(y: u32) callconv(.c) bool {
    return (y % 4 == 0 and y % 100 != 0) or (y % 400 == 0);
}

// Single entry that the runner invokes after setting `year_input`.
// Returning the predicate result keeps the call from being
// optimised to a constant.
export fn run() bool {
    return leap_year(year_input);
}
